src/gas_metering/mod.rs

```diff
@@ -155,185 +166,243 @@ impl Rules for ConstantCostRules {
-					.with_type(ValueType::I64)
-					.mutable()
-					.init_expr(Instruction::I64Const(0))
-					.build(),
-			);
-			// Inject the export entry for the gas counting global
-			let ebuilder = builder::ExportBuilder::new();
-			let global_export = ebuilder
-				.field(global)
-				.with_internal(elements::Internal::Global(gas_global_idx))
-				.build();
-			mbuilder.push_export(global_export);
-
-			let func_idx = functions_space;
-
-			// Build local gas function
-			let gas_func_sig =
-				builder::SignatureBuilder::new().with_param(ValueType::I64).build_sig();
-
-			let function = builder::FunctionBuilder::new()
-				.with_signature(gas_func_sig)
-				.body()
-				.with_instructions(func_instructions.clone())
-				.build()
-				.build();
-
-			// Inject local gas function
-			mbuilder.push_function(function);
-
-			(func_idx, func_idx + 1, cost)
+			module_info.add_global(
+				GlobalType { content_type: ValType::I64, mutable: true },
+				&wasm_encoder::ConstExpr::i64_const(0),
+			)?;
+
+			module_info.add_export(global_name, ExportKind::Global, gas_global_idx)?;
+
+			// Inject the local gas function
+			let ty = Type::Func(FuncType::new(vec![ValType::I64], vec![]));
+			module_info.add_functions(&[(ty, func.clone())])?;
+
+			(
+				// Don't inject counters to the local gas function, which is the last one as
+				// it's just added. Cost for its execution is added statically before each
+				// invocation (see `inject_counter()`).
+				functions_count,
+				functions_count + 1,
+				cost,
+			)
 		},
 	};
 
-	// We need the built the module for making injections to its blocks
-	let mut resulting_module = mbuilder.build();
-
 	let mut need_grow_counter = false;
-	let mut result = Ok(());
+	let mut error = false;
+
 	// Iterate over module sections and perform needed transformations.
 	// Indexes are needed to be fixed up in `GasMeter::External` case, as it adds an imported
 	// function, which goes to the beginning of the module's functions space.
-	'outer: for section in resulting_module.sections_mut() {
-		match section {
-			elements::Section::Code(code_section) => {
-				let injection_targets = match gas_meter {
-					GasMeter::External { .. } => code_section.bodies_mut().as_mut_slice(),
-					// Don't inject counters to the local gas function, which is the last one as
-					// it's just added. Cost for its execution is added statically before each
-					// invocation (see `inject_counter()`).
-					GasMeter::Internal { .. } => {
-						let len = code_section.bodies().len();
-						&mut code_section.bodies_mut()[..len - 1]
-					},
-				};
+	if module_info.func_bodies_count > 0 {
+		let mut code_section_builder = wasm_encoder::CodeSection::new();
+		let code_item_count = module_info.func_bodies_count;
+
+		for (func_body, is_last) in module_info
+			.code_section()?
+			.ok_or_else(|| anyhow!("no code section"))?
+			.into_iter()
+			.enumerate()
+			.map(|(index, item)| (item, index as u32 == code_item_count - 1))
```

Comment:
Do we check somewhere if the number of items in code section equals `func_bodies_count`.

Reply:
`func_bodies_count` renamed to `code_section_entry_count` for more clarity